CI: Clang CUDA

ax3l opened the pull request:
## Summary

Try to compile with vanilla Clang/LLVM's CUDA C++ frontend & its LLVM PTX backend.

Vanilla LLVM is used in a variety of tools that we want to use for research.

## Additional background

ping @ludgerpaehler

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

ax3l committed:
CI: Clang CUDA

Try to compile with vanilla Clang/LLVM's CUDA C++ frontend & its
LLVM PTX backend.

ax3l committed:
work-arounds

ax3l committed:
Clang 14 only supports CUDA <=11.5

Ref.:
https://gist.github.com/ax3l/9489132#clang--x-cuda

ax3l commented on .github/workflows/cuda.yml:
```diff
@@ -39,6 +39,38 @@ jobs:
+  tests-cuda11-clang:
+    name: ClangCUDA@12 C++17 Release [tests]
+    runs-on: ubuntu-22.04
+    env: {CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check"}
```
```suggestion
    env: {CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check -stdlib=libc++"}
```

ax3l commented on .github/workflows/cuda.yml:
```diff
@@ -39,6 +39,38 @@ jobs:
+            -DCMAKE_CUDA_COMPILER=$(which clang++)       \
+            -DCMAKE_CUDA_HOST_COMPILER=$(which clang++)  \
+            -DCMAKE_CUDA_ARCHITECTURES=70                \
+            -DAMReX_CUDA_ARCH=7.0 \
+            -D_CMAKE_CUDA_WHOLE_FLAG="-c"
```
```suggestion
            -DAMReX_CUDA_ARCH=7.0
```

ax3l commented on .github/workflows/cuda.yml:
```diff
@@ -39,6 +39,38 @@ jobs:
+            -DAMReX_FORTRAN=OFF                          \
+            -DAMReX_PARTICLES=ON                         \
+            -DAMReX_GPU_BACKEND=CUDA                     \
+            -DCMAKE_C_COMPILER=$(which clang)            \
+            -DCMAKE_CXX_COMPILER=$(which clang++)        \
+            -DCMAKE_CUDA_COMPILER=$(which clang++)       \
+            -DCMAKE_CUDA_HOST_COMPILER=$(which clang++)  \
```
```suggestion
            -DCMAKE_C_COMPILER=$(which clang-15)            \
            -DCMAKE_CXX_COMPILER=$(which clang++-15)        \
            -DCMAKE_CUDA_COMPILER=$(which clang++-15)       \
            -DCMAKE_CUDA_HOST_COMPILER=$(which clang++-15)  \
```

ax3l commented on .github/workflows/dependencies/dependencies_nvcc11_ubuntu22.sh:
```diff
@@ -0,0 +1,34 @@
+sudo apt-get install -y \
+    build-essential     \
+    ca-certificates     \
+    cmake               \
```
```suggestion
    clang-15 libc++-15-dev libc++abi-15-dev libc++1-15 libc++abi1-15 \
    cmake               \
```

ax3l committed:
Clang-15 + CTK 11.2 + libc++

ax3l commented on .github/workflows/cuda.yml:
```diff
@@ -39,6 +39,37 @@ jobs:
+  tests-cuda11-clang:
+    name: ClangCUDA@12 C++17 Release [tests]
+    runs-on: ubuntu-22.04
+    env: {CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check -stdlib=libc++"}
```
```suggestion
    env:
      CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check"
      CUDAFLAGS: "-stdlib=libc++"
```

ax3l commented on .github/workflows/cuda.yml:
```diff
@@ -58,6 +58,39 @@ jobs:
+    name: ClangCUDA@12 C++17 Release [tests]
+    runs-on: ubuntu-22.04
+    env:
+      CXXFLAGS: "--cuda-gpu-arch=sm_70 --no-cuda-version-check"
```
```suggestion
```

ax3l committed:
Cleanup & CCache

WeiqunZhang commented on .github/workflows/dependencies/dependencies_llvm_cuda11_ubuntu22.sh:
```diff
@@ -0,0 +1,36 @@
+sudo dpkg -i cuda-keyring_1.0-1_all.deb
+sudo apt-get update
+sudo apt-get install -y \
+    cuda-command-line-tools-11-2 \
+    cuda-compiler-11-2           \
+    cuda-cupti-dev-11-2          \
+    cuda-minimal-build-11-2      \
+    cuda-nvml-dev-11-2           \
+    cuda-nvtx-11-2               \
+    libcurand-dev-11-2
+sudo ln -s cuda-11.2 /usr/local/cuda
```
```suggestion
    cuda-command-line-tools-11-7 \
    cuda-compiler-11-7           \
    cuda-cupti-dev-11-7          \
    cuda-minimal-build-11-7      \
    cuda-nvml-dev-11-7           \
    cuda-nvtx-11-7               \
    libcurand-dev-11-7
sudo ln -s cuda-11.7 /usr/local/cuda
```
The oldest version of cuda on ubuntu 22.04 is 11.7.

WeiqunZhang committed:
Update .github/workflows/dependencies/dependencies_llvm_cuda11_ubuntu22.sh

WeiqunZhang commented on .github/workflows/cuda.yml:
```diff
@@ -58,6 +58,55 @@ jobs:
         ccache -s
         du -hs ~/.cache/ccache
 
+  # Build libamrex and all tests for CUDA with LLVM Clang + libc++ + CTK 11.2
```
```suggestion
  # Build libamrex and all tests for CUDA with LLVM Clang + libc++ + CTK 11.7
```

WeiqunZhang committed:
Update .github/workflows/cuda.yml

WeiqunZhang committed:
Fix syntax

WeiqunZhang commented:
There is a warning.
```
clang: warning: CUDA version is newer than the latest supported version 11.5 [-Wunknown-cuda-version]
```
We can probably ignore the warning or switch to ubuntu-20.04 so that we can use cuda <= 11.5. But there is probably no apt package for clang 15 on ubuntu-20.04.

WeiqunZhang committed:
Add link flag -stdlib=libc++

WeiqunZhang commented on .github/workflows/cuda.yml:
```diff
@@ -58,6 +58,57 @@ jobs:
+
+        cmake -S . -B build                              \
+            -DCMAKE_VERBOSE_MAKEFILE=ON                  \
+            -DAMReX_EB=ON                                \
```
```suggestion
            -DAMReX_MPI=OFF                              \
            -DAMReX_EB=ON                                \
```

WeiqunZhang committed:
Update .github/workflows/cuda.yml

Disable MPI for now, because CMake fails to find MPI due to the -std=libc++ in LDFLAGS. But we may have to use that flag, otherwise there are a bunch of errors about missing std functions at link time.

WeiqunZhang commented:
It compiles now. Ccache does not seem to work.
```
2023-06-04T22:38:36.2075501Z Cacheable calls:     0 / 223 ( 0.00%)
2023-06-04T22:38:36.2075785Z   Hits:              0
2023-06-04T22:38:36.2077008Z     Direct:          0
2023-06-04T22:38:36.2077487Z     Preprocessed:    0
2023-06-04T22:38:36.2077829Z   Misses:            0
2023-06-04T22:38:36.2078057Z Uncacheable calls: 223 / 223 (100.0%)
```

WeiqunZhang commented:
https://ccache.dev/releasenotes.html#_ccache_4_2
```
Added support for compiler option -x cuda, understood by Clang.
```
It should work since we use ccache 4.8.

ax3l commented on .github/workflows/cuda.yml:
```diff
@@ -58,6 +58,58 @@ jobs:
 
+  # Build libamrex and all tests for CUDA with LLVM Clang + libc++ + CTK 11.7
+  tests-cuda11-clang:
+    name: ClangCUDA@12 C++17 Release [tests]
```
```suggestion
    name: Clang@15 CUDA@11.7 C++17 Release [tests]
```

ax3l committed:
Update Job Name

WeiqunZhang commented:
More information on why ccache fails.
```
Cacheable calls:          0 / 107 ( 0.00%)
  Hits:                   0
    Direct:               0
    Preprocessed:         0
  Misses:                 0
Uncacheable calls:      107 / 107 (100.0%)
  Preprocessing failed: 107 / 107 (100.0%)
Successful lookups:
  Direct:                 0 / 107 ( 0.00%)
  Preprocessed:           0
Local storage:
  Cache size (GB):      0.0 / 0.2 ( 0.00%)
  Files:                  0
  Hits:                   0
  Misses:                 0
  Reads:                107
  Writes:                 0
  ```

WeiqunZhang commented:
Looks like the issue can be fixed with `CCACHE_DEPEND=1`. I now get
```
023-06-07T23:07:59.8206743Z Cacheable calls:    107 / 107 (100.0%)
2023-06-07T23:07:59.8207355Z   Hits:               0 / 107 ( 0.00%)
2023-06-07T23:07:59.8207876Z     Direct:           0
2023-06-07T23:07:59.8251756Z     Preprocessed:     0
2023-06-07T23:07:59.8257964Z   Misses:           107 / 107 (100.0%)
2023-06-07T23:07:59.8260542Z Successful lookups:
2023-06-07T23:07:59.8260953Z   Direct:             0 / 107 ( 0.00%)
2023-06-07T23:07:59.8261337Z   Preprocessed:       0
2023-06-07T23:07:59.8261697Z Local storage:
2023-06-07T23:07:59.8262068Z   Cache size (GB):  0.0 / 0.2 ( 5.33%)
2023-06-07T23:07:59.8262453Z   Files:            214
2023-06-07T23:07:59.8262795Z   Hits:               0 / 107 ( 0.00%)
2023-06-07T23:07:59.8263158Z   Misses:           107 / 107 (100.0%)
2023-06-07T23:07:59.8264383Z   Reads:            107
2023-06-07T23:07:59.8264638Z   Writes:           214
```
for the first run, and then
```
2023-06-07T23:11:49.2139440Z Cacheable calls:    107 / 107 (100.0%)
2023-06-07T23:11:49.2139706Z   Hits:             107 / 107 (100.0%)
2023-06-07T23:11:49.2139943Z     Direct:         107 / 107 (100.0%)
2023-06-07T23:11:49.2140206Z     Preprocessed:     0 / 107 ( 0.00%)
2023-06-07T23:11:49.2140455Z   Misses:             0 / 107 ( 0.00%)
2023-06-07T23:11:49.2140679Z Successful lookups:
2023-06-07T23:11:49.2140914Z   Direct:           107 / 107 (100.0%)
2023-06-07T23:11:49.2141152Z   Preprocessed:       0
2023-06-07T23:11:49.2141380Z Local storage:
2023-06-07T23:11:49.2141612Z   Cache size (GB):  0.0 / 0.2 ( 5.33%)
2023-06-07T23:11:49.2141830Z   Files:            214
2023-06-07T23:11:49.2142053Z   Hits:             107 / 107 (100.0%)
2023-06-07T23:11:49.2142285Z   Misses:             0 / 107 ( 0.00%)
2023-06-07T23:11:49.2142507Z   Reads:            214
2023-06-07T23:11:49.2142723Z   Writes:             0
```
